compiler/passes/InitErrorHandling.cpp

```diff
@@ -0,0 +1,314 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "InitErrorHandling.h"
+
+#include "AstVisitor.h"
+#include "AstVisitorTraverse.h"
+#include "ForallStmt.h"
+#include "IfExpr.h"
+#include "LoopExpr.h"
+#include "stmt.h"
+#include "astutil.h"
+
+/************************************* | **************************************
+*                                                                             *
+*                                                                             *
+*                                                                             *
+************************************** | *************************************/
+
+InitErrorHandling::InitErrorHandling(FnSymbol* fn) {
+  mFn            = fn;
+  mPhase         = startPhase(fn);
+}
+
+InitErrorHandling::InitErrorHandling(CondStmt* cond, const InitErrorHandling& curr) {
+  mFn            = curr.mFn;
+  mPhase         = curr.mPhase;
+}
+
+void InitErrorHandling::merge(const InitErrorHandling& fork) {
+  mPhase     = fork.mPhase;
+}
+
+FnSymbol* InitErrorHandling::theFn() const {
+  return mFn;
+}
+
+InitErrorHandling::InitPhase InitErrorHandling::currPhase() const {
+  return mPhase;
+}
+
+bool InitErrorHandling::isPhase0() const {
+  return mPhase == cPhase0;
+}
+
+bool InitErrorHandling::isPhase1() const {
+  return mPhase == cPhase1;
+}
+
+bool InitErrorHandling::isPhase2() const {
+  return mPhase == cPhase2;
+}
+
+/************************************* | **************************************
+*                                                                             *
+*                                                                             *
+*                                                                             *
+************************************** | *************************************/
+static const char* initName(CallExpr* stmt);
+
+bool isResolvedSuperInit(CallExpr* stmt) {
+  const char* name = initName(stmt);
+
+  return name != NULL && strcmp(name, "super") == 0 ? true : false;
+}
+
+bool isResolvedThisInit (CallExpr* stmt) {
+  const char* name = initName(stmt);
+
+  return name != NULL && strcmp(name, "this") == 0 ? true: false;
+}
+
+bool InitErrorHandling::isInitDone (CallExpr* stmt) const {
+  return stmt->isPrimitive(PRIM_INIT_DONE);
+}
+
+bool isInitStmt(CallExpr* stmt) {
+  return isResolvedSuperInit(stmt) == true || isResolvedThisInit(stmt) == true;
+}
+
+static const char* initName(CallExpr* stmt) {
+  const char* retval = NULL;
+
+  // Not in a primitive
+  if (stmt->baseExpr != NULL) {
+    if (SymExpr* base = toSymExpr(stmt->baseExpr)) {
+
+      // Call is to an initializer
+      if (strcmp(base->symbol()->name, "init") == 0 &&
+          stmt->numActuals() >= 1) {
+        SymExpr* firstArg = toSymExpr(stmt->get(1));
+        if (strcmp(firstArg->symbol()->name, "super") == 0 ||
```

Comment:
these strcmps seem a bit hacky. Could you put a comment to indicate they are a workaround for something-or-other?